tasks/insertionTasks.ts

```diff
@@ -0,0 +1,235 @@
+    console.log(`Generating PR list from ${startSHA} to ${endSHA}...`);
+
+    try {
+        const { stdout } = await execAsync(
+            `cd "${roslynRepoPath}" && roslyn-tools pr-finder -s "${startSHA}" -e "${endSHA}" --format changelog --label VSCode`,
+            { maxBuffer: 10 * 1024 * 1024 } // 10MB buffer
+        );
+        return stdout || '(failed to generate PR list, see pipeline for details)';
+    } catch (error) {
+        const errorMessage = error instanceof Error ? error.message : String(error);
+        logWarning(`PR finder failed, using empty list: ${errorMessage}`);
+        if (error instanceof Error && error.stack) {
+            console.log(`##[debug]${error.stack}`);
+        }
+        return '(failed to generate PR list, see pipeline for details)';
+    }
+}
+
+async function updatePackageJson(newVersion: string): Promise<void> {
+    console.log(`Updating package.json with Roslyn version ${newVersion}...`);
+    const packageJsonPath = 'package.json';
+    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
+
+    if (!packageJson.defaults) {
+        throw new Error('Could not find defaults section in package.json');
+    }
+    packageJson.defaults.roslyn = newVersion;
+
+    fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2) + '\n');
+}
+
+async function updateChangelog(version: string, prList: string, buildNumber?: string, buildId?: string): Promise<void> {
+    console.log('Updating CHANGELOG.md...');
+    const changelogPath = 'CHANGELOG.md';
+    const text = fs.readFileSync(changelogPath, 'utf8');
+    const NL = os.EOL;
+
+    // Prepare PR list (filter out 'View Complete Diff' lines)
+    const prLines = prList
+        ? prList
+            .split(/\r?\n/)
+            .filter((l) => l.trim() && !l.includes('View Complete Diff'))
+            .map((line) => line.trim())
+        : [];
+
+    // Format PR list as sub-items with proper indentation
+    const formattedPRList = prLines.length > 0
+        ? prLines.map((line) => {
+            // If the line already starts with *, keep it, otherwise add it
+            const cleanLine = line.startsWith('*') ? line.substring(1).trim() : line;
+            return `  * ${cleanLine}`;
+          }).join(NL)
+        : '';
+
+    // Find the first top-level header "# ..."
+    const topHeaderRegex = /^(# .*?)(\r?\n|$)/m;
+    const headerMatch = topHeaderRegex.exec(text);
+    if (!headerMatch) {
+        throw new Error('CHANGELOG.md must contain at least one top-level header (#)');
+    }
+
+    const headerEndLineIndex = headerMatch.index + headerMatch[0].length;
+
+    // Prepare new Roslyn block
+    const prLink = buildNumber && buildId
+        ? `[#${buildNumber}](https://dev.azure.com/dnceng/internal/_build/results?buildId=${buildId})`
+        : '[#TBD](TBD)';
+
+    let newRoslynBlock = `* Bump Roslyn to ${version} (PR: ${prLink})`;
```

Comment:
This doesn't seem quite right.  I do like the link to the actual build, but that should probably be part of either the PR description and doesn't need to be in the changelog itself.

The `(PR: <pr>)` part of the changelog typically includes a link to the PR in vscode-csharp that is responsible for the version bump - in this case that would be the PR we are creating.

In order to get that link however we have to create the PR first.  What we might have to do is create the PR in one step, then push to the branch again to update the PR # in the changelog (after we know what the PR number is).  We could potentially do the entire changelog update in a second commit after the PR is created.